File: blo/go/parse.go
package main

import (
	"errors"
)

type tokenStream struct {
	tokens  <-chan Token
	pending Token
}

func (t *tokenStream) next() Token {
	if t.pending.Token != "" {
		token := t.pending
		t.pending.Token = ""
		return token
	}
	if token, ok := <-t.tokens; ok {
		return token
	}
	return Token{Token: ""}
}

func (t *tokenStream) peek() Token {
	if t.pending.Token == "" {
		t.pending = t.next()
	}
	return t.pending
}

func (t *tokenStream) skipNewlines() {
	for t.peek().Token == "\n" {
		t.next()
	}
}

func Parse(tokens <-chan Token) (*Ast, error) {
	tokenStream := &tokenStream{tokens: tokens}
	ast := newAst()
	for {
		tokenStream.skipNewlines()
		switch tok := tokenStream.peek(); tok.Token {
		case "":
			return ast, nil
		case "import":
			if err := parseImport(tokenStream, ast); err != nil {
				return ast, err
			}
		case "type":
			if _, err := parseType(tokenStream, ast); err != nil {
				return ast, err
			}
		case "func":
			if funcDecl, err := parseFunc(tokenStream, ast); err != nil {
				return ast, err
			} else if stmts, err := parseStatementBlock(tokenStream); err != nil {
				return ast, err
			} else {
				funcDecl.Body = stmts
			}
		default:
			return ast, errors.New(tok.Location.String() + ": Expected 'import', 'type', or 'func', got:" + tok.Token)
		}
	}
}

func parseImport(tokenStream *tokenStream, ast *Ast) error {
	tokenStream.skipNewlines()
	if tok := tokenStream.next(); tok.Token != "import" {
		panic("Expected 'import', got:" + tok.Token)
	}
	tokenStream.skipNewlines()
	switch tok := tokenStream.peek(); tok.Token {
	case "type":
		if typeDecl, err := parseType(tokenStream, ast); err != nil {
			return err
		} else {
			typeDecl.Imported = true
		}
	case "func":
		if funcDecl, err := parseFunc(tokenStream, ast); err != nil {
			return err
		} else {
			funcDecl.Imported = true
		}
		switch tok := tokenStream.peek(); tok.Token {
		case ";", "\n":
			tokenStream.next()
		default:
			return errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
		}
	default:
		return errors.New(tok.Location.String() + ": Expected 'type' or 'func', got:" + tok.Token)
	}
	return nil
}

func parseType(tokenStream *tokenStream, ast *Ast) (*Type, error) {
	tokenStream.skipNewlines()
	typeDecl := &Type{}
	if tok := tokenStream.next(); tok.Token != "type" {
		panic("Expected 'type', got:" + tok.Token)
	} else {
		typeDecl.Location = tok.Location
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.next(); !tok.IsIdentifier() {
		return nil, errors.New(tok.Location.String() + ": Expected type name identifier, got:" + tok.Token)
	} else {
		typeDecl.Name = tok.Token
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.next(); tok.Token != "{" {
		return nil, errors.New(tok.Location.String() + ": Expected '{', got:" + tok.Token)
	}
	if _, ok := ast.Types[typeDecl.Name]; ok {
		return nil, errors.New(typeDecl.Location.String() + ": Duplicate type name:" + typeDecl.Name)
	}
	for {
		tokenStream.skipNewlines()
		tok := tokenStream.peek()
		if tok.Token == "}" {
			tokenStream.next()
			break
		} else if !tok.IsIdentifier() {
			return nil, errors.New(tok.Location.String() + ": Expected field name identifier, got:" + tok.Token)
		}
		startIndex := len(typeDecl.Fields)
		for {
			tok = tokenStream.next()
			if !tok.IsIdentifier() {
				return nil, errors.New(tok.Location.String() + ": Expected field name identifier, got:" + tok.Token)
			}
			typeDecl.Fields = append(typeDecl.Fields, &Var{Name: tok.Token})
			tok = tokenStream.peek()
			if tok.IsIdentifier() {
				for i := startIndex; i < len(typeDecl.Fields); i++ {
					typeDecl.Fields[i].TypeName = tok.Token
				}
				tokenStream.next()
				break
			} else if tok.Token == ";" || tok.Token == "\n" || tok.Token == "}" {
				break
			} else if tok.Token == "," {
				tokenStream.next()
			} else {
				return nil, errors.New(tok.Location.String() + ": Expected field type identifier, ';', or ',', got:" + tok.Token)
			}
		}
		tok = tokenStream.peek()
		if tok.Token != ";" && tok.Token != "\n" && tok.Token != "}" {
			return nil, errors.New(tok.Location.String() + ": Expected ';' or '}', got:" + tok.Token)
		}
		if tok.Token == ";" {
			tokenStream.next()
		}
		tokenStream.skipNewlines()
	}
	names := make(map[string]bool)
	for _, field := range typeDecl.Fields {
		if _, ok := names[field.Name]; ok {
			return nil, errors.New(typeDecl.Location.String() + ": Duplicate field name:" + field.Name + " for type:" + typeDecl.Name)
		}
		names[field.Name] = true
	}
	ast.Types[typeDecl.Name] = typeDecl
	return typeDecl, nil
}

func parseFunc(tokenStream *tokenStream, ast *Ast) (*Func, error) {
	tokenStream.skipNewlines()
	funcDecl := &Func{}
	if tok := tokenStream.next(); tok.Token != "func" {
		panic("Expected 'func', got:" + tok.Token)
	} else {
		funcDecl.Location = tok.Location
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.next(); !tok.IsIdentifier() {
		return nil, errors.New(tok.Location.String() + ": Expected function name identifier, got:" + tok.Token)
	} else {
		funcDecl.Name = tok.Token
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.next(); tok.Token != "(" {
		return nil, errors.New(tok.Location.String() + ": Expected '(', got:" + tok.Token)
	}

	if _, ok := ast.Funcs[funcDecl.Name]; ok {
		return nil, errors.New(funcDecl.Location.String() + ": Duplicate func name:" + funcDecl.Name)
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.peek(); tok.Token == ")" {
		tokenStream.next()
	} else {
		for {
			startIndex := len(funcDecl.Params)
			for {
				tokenStream.skipNewlines()
				tok := tokenStream.next()
				if !tok.IsIdentifier() {
					return nil, errors.New(tok.Location.String() + ": Expected parameter name identifier, got:" + tok.Token)
				}
				funcDecl.Params = append(funcDecl.Params, &Var{Name: tok.Token})
				tokenStream.skipNewlines()
				tok = tokenStream.next()
				if tok.Token == "," {
					continue
				} else if !tok.IsIdentifier() {
					return nil, errors.New(tok.Location.String() + ": Expected parameter type name identifier or ',', got:" + tok.Token)
				}
				for i := startIndex; i < len(funcDecl.Params); i++ {
					funcDecl.Params[i].TypeName = tok.Token
				}
				break
			}
			tok = tokenStream.next()
			if tok.Token == ")" {
				break
			} else if tok.Token != "," {
				return nil, errors.New(tok.Location.String() + ": Expected ',' or ')', got:" + tok.Token)
			}
		}
	}
	if tok := tokenStream.peek(); tok.IsIdentifier() {
		tokenStream.next()
		funcDecl.TypeName = tok.Token
	}
	names := make(map[string]bool)
	for _, param := range funcDecl.Params {
		if _, ok := names[param.Name]; ok {
			return nil, errors.New(funcDecl.Location.String() + ": Duplicate func parameter name:" + param.Name + " for func:" + funcDecl.Name)
		}
		names[param.Name] = true
	}
	ast.Funcs[funcDecl.Name] = funcDecl
	return funcDecl, nil
}

func parseStatementBlock(tokenStream *tokenStream) (*StmtBlock, error) {
	tokenStream.skipNewlines()
	stmtBlock := &StmtBlock{}
	if tok := tokenStream.next(); tok.Token != "{" {
		return nil, errors.New(tok.Location.String() + ": Expected '{', got:" + tok.Token)
	} else {
		stmtBlock.location = tok.Location
	}
	for {
		tokenStream.skipNewlines()
		tok := tokenStream.peek()
		switch tok.Token {
		case "}":
			tokenStream.next()
			return stmtBlock, nil
		case "{":
			if stmt, err := parseStatementBlock(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "var":
			if stmt, err := parseStmtVar(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "if":
			if stmt, err := parseStmtIf(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "for":
			if stmt, err := parseStmtFor(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "break":
			if stmt, err := parseStmtBreak(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "return":
			if stmt, err := parseStmtReturn(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		case "set", "clear":
			if stmt, err := parseStmtSetClear(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		default:
			if !tok.IsIdentifier() {
				return nil, errors.New(tok.Location.String() + ": Expected '}', 'var', 'if', 'for', 'break', 'return', 'set', 'clear', or expression, got:" + tok.Token)
			}
			if stmt, err := parseStmtExprOrAssign(tokenStream); err != nil {
				return nil, err
			} else {
				stmtBlock.Stmts = append(stmtBlock.Stmts, stmt)
			}
		}
	}
}

func parseStmtVar(tokenStream *tokenStream) (*StmtVar, error) {
	tokenStream.skipNewlines()
	stmt := &StmtVar{}
	if tok := tokenStream.next(); tok.Token != "var" {
		panic("Expected 'var', got:" + tok.Token)
	} else {
		stmt.location = tok.Location
	}
	if tok := tokenStream.next(); !tok.IsIdentifier() {
		return nil, errors.New(tok.Location.String() + ": Expected variable name identifier, got:" + tok.Token)
	} else {
		stmt.Name = tok.Token
	}
	if tok := tokenStream.next(); !tok.IsIdentifier() {
		return nil, errors.New(tok.Location.String() + ": Expected type name identifier, got:" + tok.Token)
	} else {
		stmt.TypeName = tok.Token
	}
	switch tok := tokenStream.peek(); tok.Token {
	case "=":
		tokenStream.next()
		if expr, err := parseExpr(tokenStream, false); err != nil {
			return nil, err
		} else {
			stmt.Expr = expr
		}
	case ";", "\n":
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';' or '=', got:" + tok.Token)
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
	}
	return stmt, nil
}

func parseStmtIf(tokenStream *tokenStream) (*StmtIf, error) {
	tokenStream.skipNewlines()
	stmt := &StmtIf{}
	if tok := tokenStream.next(); tok.Token != "if" {
		panic("Expected 'if', got:" + tok.Token)
	} else {
		stmt.location = tok.Location
	}
	if expr, err := parseExpr(tokenStream, true); err != nil {
		return nil, err
	} else {
		stmt.Expr = expr
	}
	if stmts, err := parseStatementBlock(tokenStream); err != nil {
		return nil, err
	} else {
		stmt.Stmts = stmts
	}
	tokenStream.skipNewlines()
	if tokenStream.peek().Token != "else" {
		return stmt, nil
	}
	tokenStream.next()
	tokenStream.skipNewlines()
	switch tok := tokenStream.peek(); tok.Token {
	case "if":
		if ifStmt, err := parseStmtIf(tokenStream); err != nil {
			return nil, err
		} else {
			stmt.ElseIf = ifStmt
		}
	case "{":
		if stmts, err := parseStatementBlock(tokenStream); err != nil {
			return nil, err
		} else {
			stmt.Else = stmts
		}
	default:
		return nil, errors.New(tok.Location.String() + ": Expected '{' or 'if', got:" + tok.Token)
	}
	return stmt, nil
}

func parseStmtFor(tokenStream *tokenStream) (*StmtFor, error) {
	tokenStream.skipNewlines()
	stmt := &StmtFor{}
	if tok := tokenStream.next(); tok.Token != "for" {
		panic("Expected 'for', got:" + tok.Token)
	} else {
		stmt.location = tok.Location
	}
	tokenStream.skipNewlines()
	if tok := tokenStream.peek(); tok.IsIdentifier() {
		tokenStream.next()
		stmt.Label = tok.Token
	}
	if stmts, err := parseStatementBlock(tokenStream); err != nil {
		return nil, err
	} else {
		stmt.Stmts = stmts
	}
	return stmt, nil
}

func parseStmtBreak(tokenStream *tokenStream) (*StmtBreak, error) {
	tokenStream.skipNewlines()
	stmt := &StmtBreak{}
	if tok := tokenStream.next(); tok.Token != "break" {
		panic("Expected 'break', got:" + tok.Token)
	} else {
		stmt.location = tok.Location
	}
	if tok := tokenStream.peek(); tok.IsIdentifier() {
		tokenStream.next()
		stmt.Label = tok.Token
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
	}
	return stmt, nil
}

func parseStmtReturn(tokenStream *tokenStream) (*StmtReturn, error) {
	tokenStream.skipNewlines()
	stmt := &StmtReturn{}
	if tok := tokenStream.next(); tok.Token != "return" {
		panic("Expected 'return', got:" + tok.Token)
	} else {
		stmt.location = tok.Location
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
	default:
		if tok.IsIdentifier() {
			if expr, err := parseExpr(tokenStream, false); err != nil {
				return nil, err
			} else {
				stmt.Expr = expr
			}
		} else {
			return nil, errors.New(tok.Location.String() + ": Expected ';' or expression, got:" + tok.Token)
		}
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
	}
	return stmt, nil
}

func parseStmtSetClear(tokenStream *tokenStream) (*StmtSetClear, error) {
	tokenStream.skipNewlines()
	stmtTok := tokenStream.next()
	if stmtTok.Token != "set" && stmtTok.Token != "clear" {
		panic("Expected 'set' or 'clear', got:" + stmtTok.Token)
	}
	tokenStream.skipNewlines()
	stmt := &StmtSetClear{location: stmtTok.Location, Value: stmtTok.Token == "set"}
	if expr, err := parseExpr(tokenStream, false); err != nil {
		return nil, err
	} else {
		stmt.Expr = expr
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
	}
	return stmt, nil
}

func parseStmtExprOrAssign(tokenStream *tokenStream) (Stmt, error) {
	tokenStream.skipNewlines()
	var lvalue Expr
	if expr, err := parseExpr(tokenStream, false); err != nil {
		return nil, err
	} else {
		lvalue = expr
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
		return &StmtExpr{Expr: lvalue}, nil
	case "=":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected '=' or ';', got:" + tok.Token)
	}
	stmt := &StmtAssign{LValue: lvalue}
	tokenStream.skipNewlines()
	if expr, err := parseExpr(tokenStream, false); err != nil {
		return nil, err
	} else {
		stmt.Expr = expr
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ";", "\n":
		tokenStream.next()
	default:
		return nil, errors.New(tok.Location.String() + ": Expected ';', got:" + tok.Token)
	}
	return stmt, nil
}

func parseExpr(tokenStream *tokenStream, ignoreNewlines bool) (Expr, error) {
	name := tokenStream.next()
	if !name.IsIdentifier() {
		return nil, errors.New(name.Location.String() + ": Expected function name identifier or local variable identifier, got:" + name.Token)
	}
	if ignoreNewlines {
		tokenStream.skipNewlines()
	}
	switch tok := tokenStream.peek(); tok.Token {
	case ".":
		return parseExprField(tokenStream, ignoreNewlines, &ExprVar{location: name.Location, Name: name.Token})
	case "(":
		if expr, err := parseExprFuncParams(tokenStream, name); err != nil {
			return nil, err
		} else {
			return parseExprField(tokenStream, ignoreNewlines, expr)
		}
	default:
		return &ExprVar{location: name.Location, Name: name.Token}, nil
	}
}

func parseExprField(tokenStream *tokenStream, ignoreNewlines bool, expr Expr) (Expr, error) {
	for {
		if ignoreNewlines {
			tokenStream.skipNewlines()
		}
		switch tok := tokenStream.peek(); tok.Token {
		case ".":
			tokenStream.next()
			tokenStream.skipNewlines()
			fieldName := tokenStream.next()
			if !fieldName.IsIdentifier() {
				return nil, errors.New(fieldName.Location.String() + ": Expected field name identifier, got:" + fieldName.Token)
			}
			expr = &ExprField{Name: fieldName.Token, Expr: expr}
		default:
			return expr, nil
		}
	}
}

func parseExprFuncParams(tokenStream *tokenStream, name Token) (*ExprFunc, error) {
	if tok := tokenStream.next(); tok.Token != "(" {
		panic("Expected '(', got:" + tok.Token)
	}
	tokenStream.skipNewlines()
	expr := &ExprFunc{location: name.Location, Name: name.Token}
	if tokenStream.peek().Token == ")" {
		tokenStream.next()
		return expr, nil
	}
	for {
		tokenStream.skipNewlines()
		if param, err := parseExpr(tokenStream, true); err != nil {
			return nil, err
		} else {
			expr.Params = append(expr.Params, param)
		}
		tokenStream.skipNewlines()
		switch tok := tokenStream.next(); tok.Token {
		case ")":
			return expr, nil
		case ",":
		default:
			return nil, errors.New(tok.Location.String() + ": Expected ')' or ',', got:" + tok.Token)
		}
	}
}

File: blo/hs/Parse.hs
module Parse
    (ParseError,SourcePos,
     Identifier(..),Definition(..),
     FuncHeader(..),TypeField(..),Var(..),
     Stmt(..),Expr(..),
     parse)
where

import Control.Monad(unless,void,when)
import Data.Set(Set,fromList,member)
import Text.ParserCombinators.Parsec
    (ParseError,Parser,SourcePos,
     anyChar,char,eof,getPosition,lookAhead,many,many1,manyTill,
     oneOf,optionMaybe,newline,noneOf,
     sepBy,skipMany,string,try,
     (<|>),(<?>))
import qualified Text.ParserCombinators.Parsec as Parsec

parse :: String -> String -> Either ParseError [Definition]
parse filename source = Parsec.parse parser filename source
  where
    parser = do
        skipNewlines
        defs <- many definition
        skipNewlines
        many (oneOf spaceChars)
        eof
        return defs

keywords :: [String]
keywords = ["type", "func", "var", "if", "else", "for", "break",
            "return", "set", "clear", "import"]

tokenChars :: String
tokenChars = "={}().,;\n"

spaceChars :: String
spaceChars = " \t\v\r" -- does not include newline

nextToken :: Parser (SourcePos,String)
nextToken = do
    many (void (oneOf spaceChars) <|> skipLineComment <|> skipBlockComment)
    pos <- getPosition
    tok <- singleCharToken <|> continueToken ""
    return (pos,tok)
  where
    singleCharToken = fmap (:[]) (oneOf tokenChars)
    continueToken pre = do
        rest <- (if pre == "" then many1 else many) (noneOf ('/' : tokenChars ++ spaceChars))
        try (throughSlash (pre ++ rest)) <|> return (pre ++ rest)
    throughSlash pre = do
        try (lookAhead (string "/*")) <|> try (lookAhead (string "//"))
                                      <|> (char '/' >> continueToken (pre ++ "/"))
    skipLineComment = do
        try (string "//")
        manyTill (noneOf "\n") newline
        return ()
    skipBlockComment = do
        try (string "/*")
        manyTill anyChar (try (string "*/"))
        return ()

skipNewlines :: Parser ()
skipNewlines = void (many (try (token "\n")))

terminateStmt :: Parser ()
terminateStmt = void (try (token ";") <|> try (token "\n")
                                      <|> try (lookAhead (token "{"))
                                      <|> try (lookAhead (token "}")))

token :: String -> Parser SourcePos
token tok = do
    (pos,t) <- nextToken
    unless (t == tok) (fail ("Expected '" ++ tok ++ "'"))
    return pos

data Identifier = Identifier SourcePos String
  deriving Show

data Definition =
    FuncDef SourcePos FuncHeader Stmt
  | FuncImport SourcePos FuncHeader
  | TypeDef SourcePos Identifier [TypeField]
  | TypeImport SourcePos Identifier [TypeField]
  deriving Show

data FuncHeader = FuncHeader Identifier [Var] (Maybe Identifier)
  deriving Show
data TypeField = TypeField Identifier (Maybe Identifier)
  deriving Show
data Var = Var Identifier Identifier
  deriving Show
data Stmt =
    StmtBlock SourcePos [Stmt]
  | StmtVar SourcePos Var (Maybe Expr)
  | StmtIf SourcePos Expr Stmt (Maybe Stmt)
  | StmtFor SourcePos (Maybe Identifier) Stmt
  | StmtBreak SourcePos (Maybe Identifier)
  | StmtReturn SourcePos (Maybe Expr)
  | StmtSetClear SourcePos Bool Expr
  | StmtAssign SourcePos Expr Expr
  | StmtExpr Expr
  deriving Show
data Expr =
    ExprVar Identifier
  | ExprFunc Identifier [Expr]
  | ExprField Expr Identifier
  deriving Show

identifier :: Parser Identifier
identifier = do
    (pos,tok) <- nextToken
    if member tok nonIdentifiers
        then fail "Expected identifier"
        else return (Identifier pos tok)
  where
    nonIdentifiers = fromList (keywords ++ map (:[]) tokenChars)

definition :: Parser Definition
definition = do
    funcDef <|> typeDef <|> importDef <?> "'func', 'type', or 'import'"
  where
    funcDef = do
        pos <- try (token "func")
        header <- funcHeader
        skipNewlines
        body <- stmtBlock
        skipNewlines
        return (FuncDef pos header body)
    typeDef = do
        pos <- try (token "type")
        skipNewlines
        name <- identifier <?> "type name"
        skipNewlines
        token "{"
        skipNewlines
        fields <- typeFields [] [] <|> return []
        token "}"
        skipNewlines
        return (TypeDef pos name fields)
    importDef = do
        pos <- try (token "import")
        skipNewlines
        funcImport pos <|> typeImport pos <?> "'func', 'type'"
    funcImport pos = do
        try (token "func")
        header <- funcHeader
        skipNewlines
        return (FuncImport pos header)
    typeImport pos = do
        try (token "type")
        skipNewlines
        name <- identifier <?> "type name"
        skipNewlines
        token "{"
        skipNewlines
        fields <- typeFields [] [] <|> return []
        token "}"
        skipNewlines
        return (TypeImport pos name fields)

funcHeader :: Parser FuncHeader
funcHeader = do
    skipNewlines
    name <- identifier <?> "func name"
    skipNewlines
    token "("
    skipNewlines
    params <- funcParams [] [] <|> return []
    token ")"
    skipNewlines
    returnType <- optionMaybe (try identifier)
    return (FuncHeader name params returnType)

funcParams :: [Var] -> [Identifier] -> Parser [Var]
funcParams pendingVars pendingVarNames = do
    varName <- try identifier <?> "parameter name"
    skipNewlines
    anotherVarWithSameType varName <|> addVarType varName
  where
    anotherVarWithSameType varName = do
        try (token ",")
        skipNewlines
        funcParams pendingVars (varName:pendingVarNames)
    addVarType varName = do
        varType <- identifier <?> "parameter type"
        skipNewlines
        let vars = map (flip Var varType) (varName:pendingVarNames) ++ pendingVars
        anotherParam vars <|> return (reverse vars)
    anotherParam vars = do
        try (token ",")
        skipNewlines
        funcParams vars []

typeFields :: [TypeField] -> [Identifier] -> Parser [TypeField]
typeFields pendingFields pendingFieldNames = do
    fieldName <- try identifier <?> "field name"
    anotherFieldWithSameType fieldName <|> addFieldType fieldName
  where
    anotherFieldWithSameType fieldName = do
        try (token ",")
        skipNewlines
        typeFields pendingFields (fieldName:pendingFieldNames)
    addFieldType fieldName = do
        fieldType <- optionMaybe (try identifier)
        terminateField
        skipNewlines
        let fields = ((map (flip TypeField fieldType) (fieldName:pendingFieldNames)) ++ pendingFields)
        typeFields fields [] <|> return (reverse fields)
    terminateField = void (try (token ";") <|> try (token "\n")
                                           <|> try (lookAhead (token "}")))

stmt :: Parser Stmt
stmt = do
    stmtBlock <|> stmtVar <|> stmtIf <|> stmtFor <|> stmtBreak
              <|> stmtReturn <|> stmtSet <|> stmtClear <|> stmtAssignOrExpr

stmtBlock :: Parser Stmt
stmtBlock = do
    pos <- try (token "{")
    skipNewlines
    stmts <- many stmt
    skipNewlines
    token "}"
    return (StmtBlock pos stmts)

stmtVar :: Parser Stmt
stmtVar = do
    pos <- try (token "var")
    varName <- identifier <?> "variable name"
    varType <- identifier <?> "variable type"
    init <- optionMaybe initializer
    terminateStmt
    return (StmtVar pos (Var varName varType) init)
  where
    initializer = do
        try (token "=")
        skipNewlines
        expr False

stmtIf :: Parser Stmt
stmtIf = do
    pos <- try (token "if")
    ex <- expr True
    stmts <- stmtBlock
    skipNewlines
    elseStmt <- optionMaybe stmtElse
    return (StmtIf pos ex stmts elseStmt)
  where
    stmtElse = do
        try (token "else")
        skipNewlines
        stmtIf <|> stmtBlock

stmtFor :: Parser Stmt
stmtFor = do
    pos <- try (token "for")
    skipNewlines
    label <- optionMaybe (try identifier)
    skipNewlines
    stmts <- stmtBlock
    return (StmtFor pos label stmts)

stmtBreak :: Parser Stmt
stmtBreak = do
    pos <- try (token "break")
    label <- optionMaybe (try identifier)
    terminateStmt
    return (StmtBreak pos label)

stmtReturn :: Parser Stmt
stmtReturn = do
    pos <- try (token "return")
    ex <- optionMaybe (expr False)
    terminateStmt
    return (StmtReturn pos ex)

stmtSet :: Parser Stmt
stmtSet = do
    pos <- try (token "set")
    skipNewlines
    ex <- expr False
    terminateStmt
    return (StmtSetClear pos True ex)

stmtClear :: Parser Stmt
stmtClear = do
    pos <- try (token "clear")
    skipNewlines
    ex <- expr False
    terminateStmt
    return (StmtSetClear pos False ex)

stmtAssignOrExpr :: Parser Stmt
stmtAssignOrExpr = do
    ex <- expr False
    stmtAssign ex <|> stmtExpr ex
  where
    stmtAssign lhs = do
        pos <- try (token "=")
        skipNewlines
        rhs <- expr False
        terminateStmt
        return (StmtAssign pos lhs rhs)
    stmtExpr ex = do
        terminateStmt
        return (StmtExpr ex)

expr :: Bool -> Parser Expr
expr ignoreNewlines = do
    when ignoreNewlines skipNewlines
    name <- try identifier <?> "variable or func name"
    when ignoreNewlines skipNewlines
    exprField (ExprVar name) <|> exprFunc name <|> return (ExprVar name)
  where
    exprField ex = do
        try (token ".")
        skipNewlines
        name <- identifier <?> "field name"
        when ignoreNewlines skipNewlines
        exprField (ExprField ex name) <|> return (ExprField ex name)
    exprFunc name = do
        try (token "(")
        args <- sepBy (expr True) (try (token ","))
        skipNewlines
        token ")"
        when ignoreNewlines skipNewlines
        exprField (ExprFunc name args) <|> return (ExprFunc name args)
